File: packages/pvm-assemblyscript/assembly/instructions/min-max.ts
import {
  OPCODE_MAX,
  OPCODE_MAX_U,
  OPCODE_MIN,
  OPCODE_MIN_U,
} from '../config'
import { InstructionContext, InstructionResult } from '../types'
import { BaseInstruction } from './base'

export class MINInstruction extends BaseInstruction {
  opcode: i32 = OPCODE_MIN
  name: string = 'MIN'

  execute(context: InstructionContext): InstructionResult {
    const registerD = this.getRegisterD(context.operands)
    const registerA = this.getRegisterA(context.operands)
    const registerB = this.getRegisterB(context.operands)
    const valueA = this.getRegisterValue(context.registers, registerA)
    const valueB = this.getRegisterValue(context.registers, registerB)

    // Convert to signed values for comparison
    const signedA = this.toSigned64(valueA)
    const signedB = this.toSigned64(valueB)
    const result = i64(signedA) < i64(signedB) ? valueA : valueB

    this.setRegisterValue(context.registers, registerD, result)

    return new InstructionResult(-1)
  }
}

export class MIN_UInstruction extends BaseInstruction {
  opcode: i32 = OPCODE_MIN_U
  name: string = 'MIN_U'

  execute(context: InstructionContext): InstructionResult {
    const registerD = this.getRegisterD(context.operands)
    const registerA = this.getRegisterA(context.operands)
    const registerB = this.getRegisterB(context.operands)
    const valueA = this.getRegisterValue(context.registers, registerA)
    const valueB = this.getRegisterValue(context.registers, registerB)
    const result = valueA < valueB ? valueA : valueB

    this.setRegisterValue(context.registers, registerD, result)

    return new InstructionResult(-1)
  }
}

export class MAXInstruction extends BaseInstruction {
  opcode: i32 = OPCODE_MAX
  name: string = 'MAX'

  execute(context: InstructionContext): InstructionResult {
    const registerD = this.getRegisterD(context.operands)
    const registerA = this.getRegisterA(context.operands)
    const registerB = this.getRegisterB(context.operands)
    const valueA = this.getRegisterValue(context.registers, registerA)
    const valueB = this.getRegisterValue(context.registers, registerB)

    // Convert to signed values for comparison
    const signedA = this.toSigned64(valueA)
    const signedB = this.toSigned64(valueB)
    const result = i64(signedA) > i64(signedB) ? valueA : valueB

    this.setRegisterValue(context.registers, registerD, result)

    return new InstructionResult(-1)
  }
}

export class MAX_UInstruction extends BaseInstruction {
  opcode: i32 = OPCODE_MAX_U
  name: string = 'MAX_U'

  execute(context: InstructionContext): InstructionResult {
    const registerD = this.getRegisterD(context.operands)
    const registerA = this.getRegisterA(context.operands)
    const registerB = this.getRegisterB(context.operands)
    const valueA = this.getRegisterValue(context.registers, registerA)
    const valueB = this.getRegisterValue(context.registers, registerB)
    const result = valueA > valueB ? valueA : valueB

    this.setRegisterValue(context.registers, registerD, result)

    return new InstructionResult(-1)
  }
}

File: packages/pvm-rust/src/instructions/min_max.rs
//! MIN, MIN_U, MAX, MAX_U (three-register: reg_D = min/max(reg_A, reg_B)).

use crate::config::{OPCODE_MAX, OPCODE_MAX_U, OPCODE_MIN, OPCODE_MIN_U};
use crate::instructions::base::{parse_three_registers, InstructionHandler};
use crate::types::{InstructionContext, InstructionResult};

fn get_register(registers: &[u64; 13], index: u8) -> u64 {
    registers.get(index as usize).copied().unwrap_or(0)
}

fn set_register(registers: &mut [u64; 13], index: u8, value: u64) {
    if (index as usize) < 13 {
        registers[index as usize] = value;
    }
}

pub struct MinInstruction;
impl MinInstruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for MinInstruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_MIN)
    }
    fn name(&self) -> &'static str {
        "MIN"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_three_registers(context.operands);
        let a = get_register(context.registers, p.register_a);
        let b = get_register(context.registers, p.register_b);
        let sa = a as i64;
        let sb = b as i64;
        let result = if sa < sb { a } else { b };
        set_register(context.registers, p.register_d, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

pub struct MinUInstruction;
impl MinUInstruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for MinUInstruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_MIN_U)
    }
    fn name(&self) -> &'static str {
        "MIN_U"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_three_registers(context.operands);
        let a = get_register(context.registers, p.register_a);
        let b = get_register(context.registers, p.register_b);
        let result = if a < b { a } else { b };
        set_register(context.registers, p.register_d, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

pub struct MaxInstruction;
impl MaxInstruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for MaxInstruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_MAX)
    }
    fn name(&self) -> &'static str {
        "MAX"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_three_registers(context.operands);
        let a = get_register(context.registers, p.register_a);
        let b = get_register(context.registers, p.register_b);
        let sa = a as i64;
        let sb = b as i64;
        let result = if sa > sb { a } else { b };
        set_register(context.registers, p.register_d, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

pub struct MaxUInstruction;
impl MaxUInstruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}
impl InstructionHandler for MaxUInstruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_MAX_U)
    }
    fn name(&self) -> &'static str {
        "MAX_U"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let p = parse_three_registers(context.operands);
        let a = get_register(context.registers, p.register_a);
        let b = get_register(context.registers, p.register_b);
        let result = if a > b { a } else { b };
        set_register(context.registers, p.register_d, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}
